src/test/groovy/com/example/demo/CarTests.groovy

```diff
@@ -0,0 +1,262 @@
+            def json = new JsonSlurper().parseText(result)
+        then: "result is an entity with an corresponding id"
+            json.id == e2.id
+            json.id > 0
+            json.id != e1.id
+        cleanup:
+            deleteTestCar(e1.id)
+            deleteTestCar(e2.id)
+            deleteTestLocation(loc.id)
+
+    }
+
+    def "Car Find by Country test"()
+    {
+        given: "three entities, two from france, one from germany"
+            def loc1 = new LocationEntity()
+                loc1.setCountry("France")
+                locationEntityRepository.save(loc1);
+            def loc2 = new LocationEntity()
+                loc2.setCountry("Germany")
+                locationEntityRepository.save(loc2);
+            def e1 = new CarEntity()
+                e1.setLocation(loc1)
+                carEntityRepository.save(e1)
+            def e2 = new CarEntity()
+                e2.setLocation(loc2)
+                carEntityRepository.save(e2)
+            def e3 = new CarEntity()
+                e3.setLocation(loc1);
+                carEntityRepository.save(e3)
+        when: "controller is called with a country"
+            def result = mockMvc.perform(get("/api/car/findByLocation/$loc1.country")).andReturn().response.contentAsString
+            def json = new JsonSlurper().parseText(result)
+        then: "result is a set of entities from that country"
+            json[0].id == e1.id
+            json[1].id != e2.id
+            json[1].id == e3.id
+        cleanup:
+            deleteTestCar(e1.id)
+            deleteTestCar(e2.id)
+            deleteTestCar(e3.id)
+            deleteTestLocation(loc1.id)
+            deleteTestLocation(loc2.id)
+    }
+
+    def "Car Not Found Test"()
+    {
+        when: "controller is called with an invalid id"
+            def result = mockMvc.perform(get("/api/car/{id}","-1")).andReturn()
+        then: "an exception is thrown by the controller"
+            result.response.status == HttpStatus.NOT_FOUND.value()
+            result.response.errorMessage == "No car with that id"
+    }
+
+    def "All locations"()
+    {
+        given: "three entities"
+            def loc = locationEntityRepository.save(new LocationEntity())
+            def e1 = new CarEntity()
+                e1.setLocation(loc)
+                carEntityRepository.save(e1)
+            def e2 = new CarEntity()
+                e2.setLocation(loc)
+                carEntityRepository.save(e2)
+            def e3 = new CarEntity()
+                e3.setLocation(loc)
+                carEntityRepository.save(e3)
+        when: "controller is called for complete list"
+            def result = mockMvc.perform(get("/api/car")).andReturn().response.contentAsString
+            def json = new JsonSlurper().parseText(result)
+        then: "Count the number of objects"
+        json[0].id == e1.id
+        json[1].id == e2.id
+        json[2].id == e3.id
```

Comment:
Again something to keep in mind, this is fine for 3 users, but tests with a lot of items can be tough to read if there's 10+ of these statements. Something like this can do the job too. `json*.sid as Set == [e1.id, e2.id, e3.id] as Set`